lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+      first_available_room = all_available_rooms.first
+      return first_available_room
+    end
+
+    def confirm_valid_dates?(start_date, end_date)
+      check_dates_validity?(start_date, end_date)
+      check_dates_order?(start_date, end_date)
+    end
+
+    def get_requested_dates(start_date, end_date)
+      check_prior_today?(start_date)
+      confirm_valid_dates?(start_date, end_date)
+      return DateRange.new(start_date, end_date)
+    end
+
+    def reserve_room(input)
+      requested_dates = get_requested_dates(input[:start_date], input[:end_date])
+
+      reservation_id = @reservations.length + 1
+      block_id = input[:block_id]
+      room = get_first_available_room(requested_dates)
+
+      reservation_data = {
+        id: reservation_id,
+        block_id: block_id,
+        room: room,
+        date_range: requested_dates
+      }
+
+      new_reservation = Reservation.new(reservation_data)
+      @reservations << new_reservation
+      return new_reservation
+    end
+
+    def confirm_valid_qty?(requested_qty)
+      check_valid_num?(requested_qty)
+      check_num_requested?(requested_qty)
+    end
+
+    def get_blocks(requested_qty, requested_dates)
+      confirm_valid_qty?(requested_qty)
+      available_rooms = find_available_rooms(requested_dates)
+      check_enough_rooms?(available_rooms, requested_qty)
+      block = available_rooms.take(requested_qty)
+      return block
+    end
+
+    def block_rooms(input)
+      requested_dates = get_requested_dates(input[:start_date], input[:end_date])
+      block_id = @blocks.length + 1
+      blocked_qty_rooms = get_blocks(input[:party], requested_dates)
+
+      block_data = {
+        id: block_id,
+        party: blocked_qty_rooms,
+        date_range: requested_dates
+      }
+
+      new_block = Block.new(block_data)
+      @blocks << new_block
+      return new_block
+    end
+
+    private
+
+    class OldStartDateError < StandardError; end
+    class InvalidDateError < StandardError; end
+    class DatesOrderError < StandardError; end
+    class NoRoomsError < StandardError; end
+    class InvalidAmountRoomsError < StandardError; end
+    class TooManyRoomsError < StandardError; end
+    class NotEnoughError < StandardError; end
+
+    def check_id(id)
+      unless id.nil? || id >= 0
+        raise ArgumentError, "ID cannot be less than zero. (got #{id})"
+      end
+    end
+
+    def check_room_num(num)
+      raise ArgumentError, "Room num cannot be less than 1 or greater than #{NUM_OF_ROOMS} (got #{num})" if num <= 0 || num > NUM_OF_ROOMS
+    end
+
+    def check_prior_today?(start_date)
+      if start_date < Date.today
+        raise OldStartDateError.new("You cannot have a start date prior to Today")
+      end
+    end
+
+    def check_dates_validity?(start_date, end_date)
+      unless (start_date.is_a?(Date) && end_date.is_a?(Date))
+        raise InvalidDateError.new("That is not a Date type")
+      end
```

Comment:
The convention in Ruby is that methods that end in a `?` are expected to return either true or false. That's not quite what these are doing - raising exceptions is similar, but distinct, and the _return value_ here isn't going to be a boolean.